Aggregate failures Integration

myronmarston opened the pull request:
This is the start of the rspec-core work for the new `aggregate_failures` feature.  The part I'm working on now is getting the aggregated failure to format how we want.  To facilitate this, I've extracted a new `ExceptionFormatter` class that is able to format exceptions according to how rspec-core typically does it for its built-in formatters.  We'll then be able to apply this to each of the aggregated failures.

More to come, but I wanted to push what I had and get the travis build going.

JonRowe commented on lib/rspec/core/formatters/exception_formatter.rb:
```diff
@@ -0,0 +1,136 @@
+  module Core
+    module Formatters
+      # @private
+      class ExceptionFormatter
```
Is `ExceptionFormatter` a good name for this? It's not a "Formatter" in RSpec's sense (even though it does do formatting), we use the term printer elsewhere for this sort of thing.

myronmarston replied:
I thought it was a good name, in spite of the fact that it's not a "Formatter" in the RSpec sense....but `ExceptionPrinter` is fine by me if you prefer it.

JonRowe replied:
I think I do just so nobody naively thinks it's a formatter in its own right :)

myronmarston replied:
Well, it is labeled `@private`, so it's not really intended to be available or visible to users...

JonRowe replied:
Yeah I know, but so are the deprecation and bisect formatters etc and they're real formatters. Thanks for making the change though, :heart:

myronmarston commented:
@JonRowe, I renamed the `ExceptionFormatter` class to `ExceptionPrinter`.  However, I realized that we do use `Formatter` for at least one case that's not a formatter in the RSpec sense: `BacktraceFormatter`.  Apart from the normal specific meaning of "formatter" in RSpec, I think "formatter" is a better term than "printer" for what that class does, and I'm tempted to keep the original name.  Thoughts?

JonRowe commented:
Personally I think we should rename `BacktraceFormatter`, I'm still heavily on the "formatter has a specific meaning" within our lingo here... How about something like `FormatException` or `ExceptionDisplay`

JonRowe commented:
Perfect :heart:

>

myronmarston commented:
OK, this PR is ready for review as well.  It builds on top of rspec/rspec-expectations#733 and integrates that feature.  The code changes here provide a couple things:

* Improved failure output from `aggregate_failures` failures.
* Lots and lots of refactoring for the exception presentation logic to support this.
* `:aggregate_failures` metadata definition.

Note that this PR can't be merged until rspec/rspec-expectations#733 is, and that one has a `@wip` cuke, which passes against this rspec-core branch but fails against rspec-core's `master` branch.  So there's a bit of a circular dependency here, at least in terms of the testing, and I'm not sure what to do about it -- see the discussion about the cuke in rspec/rspec-expectations#733 for more info.

There's also an open question around what `:aggregate_failures` metadata should do when rspec-expectations is not loaded.  My currently implementation will yield a `NoMethodError` (`aggregate_failures`) which might be OK and is the simplest, but I could see us doing something a bit nicer too...

/cc @rspec/rspec (especially @xaviershay)

xaviershay commented on spec/rspec/core_spec.rb:
```diff
@@ -218,6 +218,14 @@
     end
   end
 
+  it 'uses only one thread local variable', :run_last do
```
huh, cute.

xaviershay commented:
> There's also an open question around what :aggregate_failures metadata should do when rspec-expectations is not loaded.

I'm allowed to define arbitrary metadata, right? So in that case, not being able to "overload" `aggregate_failures` if I'm not bringing it in from somewhere else seems like a breach of contract.

Or actually, we're now defining this as a reserved word in `rspec-core` itself (if we weren't this would be a cyclic dependency I think), therefore we should probably raise something along the lines of "no plugin implements this behaviour" or something.

Not really worth a lot of effort though I don't think.

xaviershay commented:
All this presentation stuff is gnarly. Good cleanups overall.

myronmarston commented:
> Not really worth a lot of effort though I don't think.

Yep, I'm leading towards "do nothing" for now.  We can always do something later -- handling this case a bit more gracefully for users who use rspec-core but not rspec-expectations won't be a breaking API change so there's no urgency to get it in now.

> All this presentation stuff is gnarly. Good cleanups overall.

Yeah, it kinda fried my brain :(.  I still feel like it's really messy but I don't really want to put the effort into further refactorings now unless there's an obvious easy, big win someone has in mind.

JonRowe commented on lib/rspec/core.rb:
```diff
@@ -119,20 +119,13 @@ def self.configure
-  # namespace.
-  def self.thread_local_metadata
-    Thread.current[:_rspec] ||= { :shared_example_group_inclusions => [] }
+    RSpec::Support.thread_local_data[:current_example] = example
```
Oh neat :)

JonRowe commented on lib/rspec/core/formatters/exception_presenter.rb:
```diff
@@ -0,0 +1,146 @@
+
+      private
+
+        if String.method_defined?(:encoding)
```
This is repeated from elsewhere isn't it? Is it worth creating something to do this rather than repeating ourselves?

myronmarston replied:
Not sure.  I simply moved this from `notifications.rb`.  Feel free to open up a follow up PR cleaning this up after this one gets merged  :).

myronmarston commented:
OK, I think this is ready for a final review (and hopefully a merge!).  Since @xaviershay's review, I added a cuke ([viewable here](https://relishapp.com/rspec-staging/rspec-core/docs/expectation-framework-integration/aggregating-failures)), and added some improvements so that nested failure aggregation blocks format well.

Want to do a final review, @xaviershay?

I do still have to remove the [Temp] commit that pins this to the rspec-expectations branch, of course.  I'll do that after we merge that PR.

xaviershay commented:
ship it!
